datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssigner.java

```diff
@@ -5,30 +5,136 @@
+    String datastreamGroupName = partitionMetadata.getDatastreamGroup().getName();
+    Map<String, PartitionThroughputInfo> partitionInfoMap = throughputInfo.getPartitionInfoMap();
+
+    // filter out all the tasks for the current datastream group, and retain assignments in a map
+    Map<String, Set<String>> newPartitions = new HashMap<>();
+    currentAssignment.values().forEach(tasks ->
+        tasks.forEach(task -> {
+          if (task.getTaskPrefix().equals(datastreamGroupName)) {
+            Set<String> retainedPartitions = new HashSet<>(task.getPartitionsV2());
+            retainedPartitions.retainAll(partitionMetadata.getPartitions());
+            newPartitions.put(task.getId(), retainedPartitions);
+          }
+    }));
+
+    // sort the current assignment's tasks on total throughput
+    Map<String, Integer> taskThroughputMap = new HashMap<>();
+    PartitionThroughputInfo defaultPartitionInfo = new PartitionThroughputInfo(DEFAULT_KB_RATE, DEFAULT_MESSAGE_RATE, "");
+    newPartitions.forEach((task, partitions) -> {
+      int totalThroughput = partitions.stream()
+          .mapToInt(p -> partitionInfoMap.getOrDefault(p, defaultPartitionInfo).getBytesInKBRate())
+          .sum();
+      taskThroughputMap.put(task, totalThroughput);
+    });
+    ArrayList<String> sortedTasks =
+        (ArrayList<String>) newPartitions.keySet().stream()
+            .sorted(Comparator.comparing(taskThroughputMap::get))
+            .collect(Collectors.toList());
+
+    ArrayList<String> recognizedPartitions = new ArrayList<>(); // partitions with throughput info
+    ArrayList<String> unrecognizedPartitions = new ArrayList<>(); // partitions without throughput info
+    for (String partition : unassignedPartitions) {
+      if (partitionInfoMap.containsKey(partition)) {
+        recognizedPartitions.add(partition);
+      } else {
+        unrecognizedPartitions.add(partition);
+      }
+    }
+
+    // sort unassigned partitions with throughput info on throughput
+    recognizedPartitions.sort((p1, p2) -> {
+      Integer p1KBRate = partitionInfoMap.get(p1).getBytesInKBRate();
+      Integer p2KBRate = partitionInfoMap.get(p2).getBytesInKBRate();
+      return p1KBRate.compareTo(p2KBRate);
+    });
+
+    // assign partitions with throughput info one by one, by putting the heaviest partition in the lightest task
+    while (recognizedPartitions.size() > 0) {
+       String heaviestPartition = recognizedPartitions.remove(recognizedPartitions.size() - 1);
+       int heaviestPartitionThroughput = partitionInfoMap.getOrDefault(heaviestPartition, defaultPartitionInfo)
+           .getBytesInKBRate();
+       String lightestTask = sortedTasks.remove(0);
+       newPartitions.get(lightestTask).add(heaviestPartition);
+       taskThroughputMap.put(lightestTask, taskThroughputMap.get(lightestTask) + heaviestPartitionThroughput);
+       this.insertTaskIntoSortedList(lightestTask, sortedTasks, taskThroughputMap);
```

Comment:
this looks like a good case of using a priority queue. should we use that instead of maintaining a similar implementation?

Reply:
You're right. Using a priority queue will eliminate the need for a binary search.